src/app/treelist-checkbox/treelist-checkbox.component.ts

```diff
@@ -0,0 +1,250 @@
+    });
+  }
+
+  public ngAfterViewInit(): void {
+    this.fillStatesEvent.subscribe(selected => {
+      this.allStudies = selected;
+      this.updateFilterStates(this.data);
+    });
+  }
+
+  public updateFilters($event): void {
+    const datasetNode = this.findNodeById(this.data, $event.target.id);
+    if ($event.target.checked) {
+      if (datasetNode.children.length > 0) {
+        this.getAllChildren(datasetNode.children, new Set(), true).forEach(child => {
+          this.allStudies.add(child);
+        });
+        this.allStudies.add(datasetNode.dataset.name);
+      } else {
+        this.allStudies.add(datasetNode.dataset.name);
+        this.updateParents(datasetNode);
+      }
+    } else if (datasetNode?.children.length > 0) {
+      this.getAllChildren(datasetNode.children, new Set(), true).forEach(child => {
+        this.changeCheckStatusEvent.emit($event.target.id);
+        this.allStudies.delete(child);
+      });
+      this.allStudies.delete($event.target.id);
+    } else if (this.allStudies.has($event.target.id)) {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.delete($event.target.id);
+    }
+    if (this.allStudies.has($event.target.id)) {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.add($event.target.id);
+    } else {
+      this.changeCheckStatusEvent.emit($event.target.id);
+      this.allStudies.delete($event.target.id);
+    }
+    this.updateFilterStates(this.data);
+  }
+
+  public updateParents(datasetNode): void {
+    const parents = this.getAllParents(datasetNode);
+    parents.forEach((parent: string) => {
+      if (parent === this.data.dataset.name) {
+        return;
+      }
+      let flag = 0;
+      this.findNodeById(this.data, parent)?.children?.forEach(child => {
+        if (this.allStudies.has(child.dataset.name)) {
+          flag += 1;
+        }
+      });
+      if (flag === parents.length) {
+        flag = 2;
+      } else if (flag !== 0 && flag < parents.length) {
+        flag = 1;
+      }
+      switch (flag) {
+        case 0: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = false;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
+        case 1: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = true;
+          });
+          break;
+        }
+        case 2: {
+          const elements = document.querySelectorAll(`[id='${parent}']`);
+          elements.forEach(element => {
+            (element as HTMLInputElement).checked = true;
+            (element as HTMLInputElement).indeterminate = false;
+          });
+          break;
+        }
```

Comment:
Repeating code - you could reduce it using ternary operators when assigning the checked and indeterminate values of the checkbox element.